pkg/segment/set.go

```diff
@@ -932,11 +981,197 @@ func (n *node) rebalanceAfterRemove(gap GapIterator) GapIterator {
+		if p.maxGap > oldMax {
+			// p and its ancestors still contain a larger gap.
+			break
+		}
+		// If new max is smaller than the old maxGap, and this gap used
+		// to be the maxGap of its parent, iterate parent's children
+		// and calculate parent's new maxGap.(It's probable that parent
+		// has two children with the old maxGap, but we need to check it anyway.)
+		parentNewMax := p.calculateMaxGapInternal()
+		if p.maxGap == parentNewMax {
+			// p and its ancestors still contain a gap of at least equal size.
+			break
+		}
+		// If p's new maxGap differs from the old one, propagate this update.
+		p.maxGap = parentNewMax
+	}
+}
+
+// updateMaxGapInternal updates the maxGap on internal nodes recursively.
+//
+// Precondition: Should only be called by updateMaxGapLeaf or updateMaxGapInternal.
+func (n *node) updateMaxGapInternal(newMaxGap Key) {
+	if !n.hasChildren {
+		panic(fmt.Sprintf("updateMaxGapInternal should always be called on non-leaf node: %v", n))
+	}
+	if newMaxGap == n.maxGap {
+		// If newMaxGap equals the old maxGap, no update is needed.
+		return
+	}
+	oldMax := n.maxGap
+	n.maxGap = newMaxGap
+	if newMaxGap > oldMax {
+		// Grow ancestor maxGaps.
+		for p := n.parent; p != nil; p = p.parent {
+			if p.maxGap >= newMaxGap {
+				// p and its ancestors already contain an equal or larger gap.
+				break
+			}
+			// Only if new maxGap is larger than parent's
+			// old maxGap, propagate this update to parent.
+			p.maxGap = newMaxGap
+		}
+		return
+	}
+	// Shrink ancestor maxGaps.
+	for p := n.parent; p != nil; p = p.parent {
+		if p.maxGap > oldMax {
+			// p and its ancestors still contain a larger gap.
+			break
+		}
+		// If new max is smaller than the old maxGap, and this gap used
+		// to be the maxGap of its parent, iterate parent's children
+		// and calculate parent's new maxGap.(It's probable that parent
+		// has two children with the old maxGap, but we need to check it anyway.)
+		parentNewMax := p.calculateMaxGapInternal()
+		if p.maxGap == parentNewMax {
+			// p and its ancestors still contain a gap of at least equal size.
+			break
+		}
+		// If p's new maxGap differs from the old one, propagate this update.
+		p.maxGap = parentNewMax
+	}
+}
+
+// updateMaxGapLocal updates maxGap of the calling node solely with no propagation to ancestor nodes.
+func (n *node) updateMaxGapLocal() {
+	var max Key
+	if !n.hasChildren {
+		// Leaf node iterates its gaps.
+		max = n.calculateMaxGapLeaf()
+	} else {
+		// Non-leaf node iterates its children.
+		max = n.children[0].maxGap
+		for i := 1; i <= n.nrSegments; i++ {
+			if current := n.children[i].maxGap; current > max {
+				max = current
+			}
+		}
+	}
+	n.maxGap = max
+}
+
+// calculateMaxGapLeaf iterates the gaps within a leaf node and calculate the max.
+//
+// Preconditions: n must be a leaf node.
+func (n *node) calculateMaxGapLeaf() Key {
+	max := GapIterator{n, 0}.Range().Length()
+	for i := 1; i <= n.nrSegments; i++ {
```

Comment:
`calculateMaxGapLeaf` may be called on an empty root, and the loop condition is off by 1 due to using `<` instead of `<=` (the use in `calculateMaxGapInternal` is correct since the number of children is `nrSegments+1`). So this should be:

```
max := Functions{}.MaxKey() - Functions{}.MinKey()
for i := 0; i < n.nrSegments; i++ {
```

Reply:
max := GapIterator{n, 0}.Range().Length() will return MaxKey() - MinKey() if it's an empty root node and the for loop will not be entered because i start at 1 but nrsegments = 0 in this case. So MaxKey() - MInKey() will be returned as expected.
This has been tested to be working properly in Remove related testcases as well